clusterloader2/pkg/measurement/common/watch_list_latency_prometheus.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/prometheus/common/model"
+
+	"k8s.io/klog/v2"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement"
+	measurementutil "k8s.io/perf-tests/clusterloader2/pkg/measurement/util"
+	"k8s.io/perf-tests/clusterloader2/pkg/util"
+)
+
+const (
+	watchListLatencyPrometheusMeasurementName = "WatchListLatencyPrometheus"
+
+	// watchListLatencyQuery placeholders must be replaced with (1) quantile (2) query window size
+	watchListLatencyQuery = "histogram_quantile(%.2f, sum(rate(apiserver_watch_list_duration_seconds{}[%v])) by (group, version, resource, scope, le))"
+)
```

Comment:
Actually - I started to wonder more generically - why can't we just reuse that other measurement that I linked?

I think we're effectively reimplementing the exact same logic and the only differences that we have are:
(1) we're using a different metric name
(2) the verb is always LIST

I think it should be possible to slightly refactor that other measurement and simply register two measurements there:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/slos/api_responsiveness_prometheus.go#L81C3-L82C1

Reply:
The `ApiResponsivenessGatherer` differs in a few places. 

First of all, it has two different modes for getting the latency metrics (simple and extended). 
In addition to gathering the latency, it collects two additional metrics: count and countFast. 
The internal data structures hold all three metrics. 
Once the metrics are collected, it supports reading a custom threshold from the config, which is used for further validation. 

I think that the refactoring would boil down to creating "generic simple latency metrics," which could potentially be reused by both implementations. 

Given that the internal data structures differ, the existing implementation would have to incorporate the generic latency metric and extend it.

Is this what you had in mind ?

Reply:
> Don't we need "_bucket" at the end of metric name?

Yes you should use the buckets with `histogram_quantile`.

Reply:
@p0lyn0mial - I'm a bit lost in your comment, so let me try to explain a bit deeper what I had in mind:

1. yes, there are two modes (simple and "normal", but the difference between these two is only how we're sampling the metrics. To be more specific, this is the only difference between these two modes:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/slos/api_responsiveness_prometheus.go#L176-L204

2. From e2e user perspective, if I want to list my objects, it doesn't really matter if the server underneath is using the list method or the new watchlist protocol. I care about the latency of getting the result

3. Because of (2), we generally don't want to introduce a separate measurement, it should actually be part of exactly the same measurement (same config, same threshold, ....)
Although, initially we may want to split that a bit for debuggability reasons.

4. So the way I think about what we should do is effectively:
- don't introduce new measurements at all - we will just slightly modify the existing one
- introduce queries that will provide us way to get sample
- in places where we gather them now, gather also the ones from watchlist:
- https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/slos/api_responsiveness_prometheus.go#L176-L204
[I'm fine with starting only with simple query to show how it will work]
- for now, just setting the "verb=watchlist" for those samples
- adding appropriate count queries too
- letting the existing logic to handle all of that as is [just separately for watchlist verb for now]

Once we prove that, we should actually merge the Samples for list & watchlist together, but let's do that as a follow-up and just start by treating them as separate things.

Reply:
OK, I think I understand, the new metric (watchlist) will end up on being reported as part of `LoadResponsiveness_PrometheusSimple` for all jobs! I like it. Thanks.

Reply:
created https://github.com/kubernetes/perf-tests/pull/2764